x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/mapper/ShapeIndexer.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.spatial.index.mapper;
+
+import org.apache.lucene.document.XYShape;
+import org.apache.lucene.geo.XYLine;
+import org.apache.lucene.geo.XYPolygon;
+import org.apache.lucene.index.IndexableField;
+import org.elasticsearch.geo.geometry.Circle;
+import org.elasticsearch.geo.geometry.Geometry;
+import org.elasticsearch.geo.geometry.GeometryCollection;
+import org.elasticsearch.geo.geometry.GeometryVisitor;
+import org.elasticsearch.geo.geometry.LinearRing;
+import org.elasticsearch.geo.geometry.MultiLine;
+import org.elasticsearch.geo.geometry.MultiPoint;
+import org.elasticsearch.geo.geometry.MultiPolygon;
+import org.elasticsearch.geo.geometry.Point;
+import org.elasticsearch.index.mapper.AbstractGeometryFieldMapper;
+import org.elasticsearch.index.mapper.ParseContext;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class ShapeIndexer implements AbstractGeometryFieldMapper.Indexer<Geometry, Geometry> {
+    private final String name;
+
+    public ShapeIndexer(String name) {
+        this.name = name;
+    }
+
+    @Override
+    public Geometry prepareForIndexing(Geometry geometry) {
+        return geometry;
+    }
+
+    @Override
+    public Class<Geometry> processedClass() {
+        return Geometry.class;
+    }
+
+    @Override
+    public List<IndexableField> indexShape(ParseContext context, Geometry shape) {
+        LuceneGeometryVisitor visitor = new LuceneGeometryVisitor(name);
+        shape.visit(visitor);
+        return visitor.fields;
+    }
+
+    private class LuceneGeometryVisitor implements GeometryVisitor<Void, RuntimeException> {
+        private List<IndexableField> fields = new ArrayList<>();
+        private String name;
+
+        private LuceneGeometryVisitor(String name) {
+            this.name = name;
+        }
+
+        @Override
+        public Void visit(Circle circle) {
+            throw new IllegalArgumentException("invalid shape type found [Circle] while indexing shape");
+        }
+
+        @Override
+        public Void visit(GeometryCollection<?> collection) {
+            for (Geometry geometry : collection) {
+                geometry.visit(this);
+            }
+            return null;
+        }
+
+        @Override
+        public Void visit(org.elasticsearch.geo.geometry.Line line) {
+            float[] x = new float[line.length()];
+            float[] y = new float[x.length];
+            for (int i = 0; i < x.length; ++i) {
+                x[i] = (float)line.getLon(i);
+                y[i] = (float)line.getLat(i);
+            }
```

Comment:
I think this can be extracted into something like `float[][] lineToFloatArray(Line line)` since it appears 3 times in this file.

Reply:
:+1: